tools/x-repo-types/runner/typeAnalyzer.go

```diff
@@ -0,0 +1,459 @@
+// Copyright (C) 2019-2023 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package main
+
+import (
+	"fmt"
+	"reflect"
+	"regexp"
+	"sort"
+	"strings"
+	"unicode"
+)
+
+// diffExclusions is a set of types to exclude. The string keys are assumed to be generated by
+// the `Type.String()` method, and these values are ignored when building up the diff.
+// These exclusions represent types that are known to serialize in the same way, but are
+// would not do so without special treatment.
+// For example, `MicroAlgos` are a struct in go-algorand but a uint64 in go-algorand-sdk.
+var diffExclusions = map[string]bool{
+	`github.com/algorand/go-algorand/data/basics :: "basics.MicroAlgos" (struct)`: true,
+}
+
+// --------------- TYPE TREE DATA STRUCTURES --------------- //
+
+// Type wraps reflect.Type and reflect.Kind to make it easier to
+// build a tree of types.
+type Type struct {
+	Depth    int
+	Type     reflect.Type
+	Kind     reflect.Kind
+	Edges    []Edge
+	children *Children
+}
```

Comment:
nit: not sure how to best choose between CS terms and domain terms. I see that you're using a tree/graph to represent the struct, which makes perfect sense, but the type names don't really help you see what is happening. The top level could have also been named `Node`, so it's a  little inconsistent with the rest of the names.

I would consider:
`Type` -> `Struct`
`Children` -> `Fields`
`Edges` -> `Tags` (or `Alias`?)

It's also fine as-is for this PR.

Reply:
# Thread on Naming
Naming is very important so I do want to find better names. Some thoughts:
1. `Children` $\rarr$ `Fields` isn't ideal since we also need to handle other sorts of nodes suchs as maps, slices, etc. Maybe `ChildTypes` ?
2. `Type` $\rarr `Struct` (for the top level name of the type). Similarly, we need to handle more than just structs. How about `TypeNode` ?
3. `Edge` $\rarr$ `Tags/Alias`. If we keep `Children`, since each edge is really naming a child how about `ChildName` or `TypeName`?
4. This suggests for the field `Edges` $\rarr$ `ChildNames` or `TypeNames` ?
5. Back to the top level how about `Type` $\rarr$ `TypeNode` ?
6. Continuing to other naming issues. Should the `runner` directory be renamed? To`typeAnalyzer`? (It's not the first camel-case directory of the repo. Cf. `config/defaultsGenerator`
7.

Reply:
I like all your suggestions better than the ones I made. :+1: 

I don't have an opinion about the directory package

Reply:
e535a8853471585ab932ea200a814225a000057b
bfd07d21cbdc90175843c28deb786e20d27b52e5
03c1cd6302cc998750480544e20b82f87866063b